ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+func (ao *onlineAccounts) onlineTotalsImpl(rnd basics.Round) (basics.MicroAlgos, error) {
+	offset, err := ao.roundParamsOffset(rnd)
+	if err != nil {
+		return basics.MicroAlgos{}, err
+	}
+
+	onlineRoundParams := ao.onlineRoundParamsData[offset]
+	return basics.MicroAlgos{Raw: onlineRoundParams.OnlineSupply}, nil
+}
+
+// LookupOnlineAccountData returns the online account data for a given address at a given round.
+func (ao *onlineAccounts) LookupOnlineAccountData(rnd basics.Round, addr basics.Address) (data basics.OnlineAccountData, err error) {
+	oad, err := ao.lookupOnlineAccountData(rnd, addr)
+	if err != nil {
+		return
+	}
+
+	data.MicroAlgosWithRewards = oad.MicroAlgosWithRewards
+	data.VotingData.VoteID = oad.VotingData.VoteID
+	data.VotingData.SelectionID = oad.VotingData.SelectionID
+	data.VotingData.StateProofID = oad.VotingData.StateProofID
+	data.VotingData.VoteFirstValid = oad.VotingData.VoteFirstValid
+	data.VotingData.VoteLastValid = oad.VotingData.VoteLastValid
+	data.VotingData.VoteKeyDilution = oad.VotingData.VoteKeyDilution
+
+	return
+}
+
+// roundOffset calculates the offset of the given round compared to the current dbRound. Requires that the lock would be taken.
+func (ao *onlineAccounts) roundOffset(rnd basics.Round) (offset uint64, err error) {
+	if rnd < ao.cachedDBRoundOnline {
+		err = &RoundOffsetError{
+			round:   rnd,
+			dbRound: ao.cachedDBRoundOnline,
+		}
+		return
+	}
+
+	off := uint64(rnd - ao.cachedDBRoundOnline)
+	if off > uint64(len(ao.deltas)) {
+		err = fmt.Errorf("round %d too high: dbRound %d, deltas %d", rnd, ao.cachedDBRoundOnline, len(ao.deltas))
+		return
+	}
+
+	return off, nil
+}
+
+// roundParamsOffset calculates the offset of the given round compared to the onlineRoundParams cache. Requires that the lock would be taken.
+func (ao *onlineAccounts) roundParamsOffset(rnd basics.Round) (offset uint64, err error) {
+	// the invariant is that the last element of ao.onlineRoundParamsData is for round ao.latest()
+	startRound := ao.latest() + 1 - basics.Round(len(ao.onlineRoundParamsData))
+	if rnd < startRound {
+		err = &RoundOffsetError{
+			round:   rnd,
+			dbRound: startRound,
+		}
+		return
+	}
+
+	off := uint64(rnd - startRound)
+	if off >= uint64(len(ao.onlineRoundParamsData)) {
+		err = fmt.Errorf("round %d too high: dbRound %d, onlineRoundParamsData %d", rnd, startRound, len(ao.onlineRoundParamsData))
+		return
+	}
+
+	return off, nil
+}
+
+// lookupOnlineAccountData returns the online account data for a given address at a given round.
+func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.Address) (data ledgercore.OnlineAccountData, err error) {
+	ao.accountsMu.RLock()
+	needUnlock := true
+	defer func() {
+		if needUnlock {
+			ao.accountsMu.RUnlock()
+		}
+	}()
+	var offset uint64
+	var paramsOffset uint64
+	var rewardsProto config.ConsensusParams
+	var rewardsLevel uint64
+	var persistedData persistedOnlineAccountData
+
+	for {
```

Comment:
Could you add a comment explaining why there is an infinite loop here?

It's not clear to me what changes the second time we go through the loop? Does it need an expiration or context to terminate early?

Reply:
^ ditto, was also confused by this logic

Reply:
We have infinite loops in all the lookup methods in acctUpdates (lookupAccount, lookupResource, lookupLatest, etc). We can definitely add more comments.